src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpTrailersHelper.cs

```diff
@@ -0,0 +1,70 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Runtime.InteropServices;
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpTrailersHelper
+    {
+#if !NETSTANDARD2_1 && !UNITTEST // The latter is required for the WinHttpHandler.Unit.Tests, which includes the source file.
+        private const string RequestMessagePropertyName = "__ResponseTrailers";
+        private class HttpResponseTrailers : HttpHeaders
+        {
+        }
+#endif
+        private static Lazy<bool> s_trailersSupported = new Lazy<bool>(GetTrailersSupported);
+        public static bool OsSupportsTrailers => s_trailersSupported.Value;
+
+        public static HttpHeaders GetResponseTrailers(HttpResponseMessage response)
+        {
+#if NETSTANDARD2_1 || UNITTEST
+            return response.TrailingHeaders;
+#else
+            HttpResponseTrailers responseTrailers = new HttpResponseTrailers();
+            response.RequestMessage.Properties[RequestMessagePropertyName] = responseTrailers;
+            return responseTrailers;
+#endif
+        }
+
+        // There is no way to verify if WINHTTP_QUERY_FLAG_TRAILERS is supported by the OS without creating a request.
+        // Instead, the WinHTTP team recommended to check if WINHTTP_OPTION_STREAM_ERROR_CODE is recognized by the OS.
+        // Both features were introduced in Manganese and are planned to be backported to older Windows versions together.
+        private static bool GetTrailersSupported()
+        {
+            SafeWinHttpHandle sessionHandle = null;
+
+            try
+            {
+                sessionHandle = Interop.WinHttp.WinHttpOpen(
```

Comment:
nit: using SafeWinHttpHandle sessionHandle = Interop.WinHttp.WinHttpOpen(

Removes the need for the try/finally.